### ISSUE TITLE
Failure to find dependency that was installed from extra index url

### ISSUE THREAD

**Issue by cornelius-braun** [trailofbits/pip-audit#599]
## Bug description

I created a requirements file for my project using `pip-compile`. To get the correct version, I added an extra url for the `torch` installation, resulting in the following command:
```
pip-compile --extra-index-url  https://download.pytorch.org/whl/cpu
```
This gives me the following `requirements.txt`
```
filelock==3.12.0
jinja2==3.1.2
markupsafe==2.1.2
mpmath==1.3.0
networkx==3.1
sympy==1.11.1
torch==2.0.0
typing-extensions==4.5.0
```

When I run `pip-audit` on this, I get the issue that `torch` is skipped from the auditing:
```
No known vulnerabilities found
Name  Skip Reason
----- ------------------------------------------------------------------------
torch Dependency not found on PyPI and could not be audited: torch (2.0.0+cpu)
```
Is this a bug or am I misusing `pip-audit`?

## Reproduction steps

I generated my requirements using
```
pip-compile --extra-index-url  https://download.pytorch.org/whl/cpu
pip-sync
```
Then I ran 
```
pip-audit -r requirements.txt --extra-index-url https://download.pytorch.org/whl/cpu
```

## Expected behavior

Auditing of all packages including `torch`.

## Screenshots and logs

The logs are as follows:

```
DEBUG:pip_audit._cli:Auditing filelock (3.12.0)
DEBUG:pip_audit._cli:Auditing Jinja2 (3.1.2)
DEBUG:pip_audit._cli:Auditing MarkupSafe (2.1.2)
DEBUG:pip_audit._cli:Auditing mpmath (1.3.0)
DEBUG:pip_audit._cli:Auditing networkx (3.1)
DEBUG:pip_audit._cli:Auditing sympy (1.11.1)
DEBUG:pip_audit._service.pypi:Dependency not found on PyPI and could not be audited: torch (2.0.0+cpu)
DEBUG:pip_audit._cli:Auditing typing_extensions (4.5.0)
No known vulnerabilities found
Name  Skip Reason
----- ------------------------------------------------------------------------
torch Dependency not found on PyPI and could not be audited: torch (2.0.0+cpu)
```

## Platform information

* OS name and version: Ubuntu 20.04.4 LTS (Focal Fossa)
* `pip-audit` version (`pip-audit -V`): 2.5.4
* Python version (`python -V` or `python3 -V`): 3.11
* `pip` version (`pip -V` or `pip3 -V`): 23.1

**Comment by cornelius-braun** [comment]
Thank you both for your elaborate replies!

> @cornelius-braun, I'm curious, when you saw the "Dependency not found on PyPI and could not be audited", was this clear enough? Is there more we could do here to say "you installed something we've never seen before, we have no way to tell you if there are known vulnerabilities for it?"

To me, it was clear that you could not find information about the torch installation because it was not found on PyPi. 

Since an `--extra-index-url` flag is supported, I was not sure, however, if you were checking for other vulnerability sources as well, as this permits to install packages from outside of PyPi. 

Based on your explanations, your procedure now makes complete sense to me.

**Comment by di** [comment]
I think we do want to support this eventually, but we could make it more clear that it's not currently supported.

**Comment by woodruffw** [comment]
Agreed! I think we can improve the user experience here with the following: 

If the user passes `--index-url` or `--extra-index-url`, we should emit a warning telling them that the PyPI vulnerability source won't necessarily report vulnerabilities for dependencies resolved from their sources.

**Comment by woodruffw** [comment]
Assigned to both myself and @tnytown, we'll triage it based on availability during the sprints.

**Comment by tufanalbayrak** [comment]
Hi. I also have the same problem. pip-audit fails to find cpu versions of torch and torchvision on PyPI. Is there any progress here? Thanks.

**Comment by woodruffw** [comment]
> Hi. I also have the same problem. pip-audit fails to find cpu versions of torch and torchvision on PyPI. Is there any progress here? Thanks.

That sounds like a different issue, since this issue is about third party index URL handling. Could you please file a separate issue and include an example for us to reproduce your problem with?

**Comment by a-recknagel** [comment]
Similar issue here, but in my case it's a private package that doesn't appear on the global pypi at all. And if it did, I'd have to treat it as a potential MIDM attack. Same as the other poster, I assumed that setting `--index-url` to the index where I downloaded the package from would solve the issue, which it didn't. Standardizing the vulnerability API would solve my problem.

**Comment by woodruffw** [comment]
> Standardizing the vulnerability API would solve my problem.

FWIW, this is not something `pip-audit` can do unilaterally: the PEP 503 and PEP 691 index APIs currently have no way to report vulnerability information, so someone would need to write a new PEP that adds vulnerability metadata to one (or both) of them.

At the moment, we rely on a PyPI-specific API ([documented here](https://docs.pypi.org/api/json/#known-vulnerabilities)) to retrieve vulnerability information. 

Another route forwards here would be to allow people to provide their own OSV-compatible API; #805 and #810 would enable that. This would be easier from a standards perspective since it isn't tied to the index APIs. I'm curious if that would be suitable for your use case?

**Comment by a-recknagel** [comment]
Thanks for the links. Yeah, I was reading the discuss thread trying to understand the amount of work, looks like it'd be a good while off, if ever. Our index is hosted by gitlab, and they're usually not the fastest in implementing features, so a solution where the vulnerability info is hosted somewhere that is not necessarily the index itself would be the most useful. In order to know which kind of API would serve best, I'd need to do some reading first, but I imagine that either works.
